Do what's needed for a new version scheme and set the next version to 3.0.0

levitte opened the pull request:
Ref: https://www.openssl.org/blog/blog/2018/11/28/version/

levitte committed:
util/mkdef.pl: prepare for DEPRECATEDIN_X

This is in preparation for new versioning scheme, where the
recommendation is to start deprecations at major version boundary.

levitte committed:
Switch future deprecation version from 1.2.0 to 3.0

This is in preparation for a switch to MAJOR.MINOR.PATCH versioning
and calling the next major version 3.0.0.

levitte committed:
Switch to MAJOR.MINOR.PATCH versioning and version 3.0.0-dev

We're strictly use version numbers of the form MAJOR.MINOR.PATCH.
Letter releases are things of days past.

The most central change is that we now express the version number with
three macros, one for each part of the version number:

    OPENSSL_VERSION_MAJOR
    OPENSSL_VERSION_MINOR
    OPENSSL_VERSION_PATCH

We also provide two additional macros to express pre-release and build
metadata information (also specified in semantic versioning):

    OPENSSL_VERSION_PRE_RELEASE
    OPENSSL_VERSION_BUILD_METADATA

To get the library's idea of all those values, we introduce the
following functions:

    unsigned int OPENSSL_version_major(void);
    unsigned int OPENSSL_version_minor(void);
    unsigned int OPENSSL_version_patch(void);
    const char *OPENSSL_version_pre_release(void);
    const char *OPENSSL_version_build_metadata(void);

Additionally, for shared library versioning (which is out of scope in
semantic versioning, but that we still need):

    OPENSSL_SHLIB_VERSION

We also provide a macro that contains the release date.  This is not
part of the version number, but is extra information that we want to
be able to display:

    OPENSSL_RELEASE_DATE

Finally, also provide the following convenience functions:

    const char *OPENSSL_version_text(void);
    const char *OPENSSL_version_text_full(void);

The following macros and functions are deprecated, and while currently
existing for backward compatibility, they are expected to disappear:

    OPENSSL_VERSION_NUMBER
    OPENSSL_VERSION_TEXT
    OPENSSL_VERSION
    OpenSSL_version_num()
    OpenSSL_version()

Also, this function is introduced to replace OpenSSL_version() for all
indexes except for OPENSSL_VERSION:

    OPENSSL_info()

For configuration, the option 'newversion-only' is added to disable all
the macros and functions that are mentioned as deprecated above.

levitte committed:
Configure: allow building without things deprecated up to and including v3.0

levitte committed:
include/openssl/opensslconf.h.in: allow future deprecation for v4.0

kroeckx commented on include/openssl/opensslv.h:
```diff
@@ -10,92 +10,126 @@
+unsigned int OPENSSL_version_minor(void);
+unsigned int OPENSSL_version_patch(void);
+const char *OPENSSL_version_pre_release(void);
+const char *OPENSSL_version_build_metadata(void);
```
Documentation

levitte replied:
Yup, I just fixed the manual

kroeckx commented on include/openssl/opensslv.h:
```diff
@@ -10,92 +10,126 @@
+    (long)( (OPENSSL_VERSION_MAJOR<<28)  \
+            |(OPENSSL_VERSION_MINOR<<20) \
+            |(OPENSSL_VERSION_PATCH<<4)  \
+            |0x0 )
```
So I understand that this would turn it into 0x30000000, which means that maximum version we can go to is 15. Maybe we should consider something else, but it should clearly be larger than the current number, or deprecating this define.

kroeckx replied:
Or we could go for a 64 bit version number ...

t-j-h replied:
Correct. OPENSSL_VERSION_NUMBER is considered deprecated usage.

kroeckx replied:
Oh, I didn't see the if OPENSSL_API_COMPAT < 0x40000000L part.

kroeckx commented on include/openssl/opensslv.h:
```diff
@@ -10,92 +10,126 @@
+ * be related to the API version expressed with the macros above.
+ * This is defined in free form.
+ */
+# define OPENSSL_SHLIB_VERSION 2
```
I would actually expect this to be the same as the MAJOR, I don't see a reason to have it different.

kroeckx commented:
We have things like -DOPENSSL_API_COMPAT=0x10100000L documented in the CHANGES file, so we might want to deprecate that itself too.

levitte commented:
Personally, I would rather get rid of the version-bound deprecation stuff *entirely* and simply have a macro `DEPRECATED` for the things we're deprecation *now*, and possibly `FUTURE_DEPRECATION` for things that are to be deprecated starting next release (and it's up to us to look for those and change them to `DEPRECATED` at each major release).

levitte commented:
Regarding the handling `OPENSSL_VERSION_PRE_RELEASE`, I'm making it undefined when there is no pre-release information so that those who wish to detect that they're building against a pre-release can do so easily with

``` C
#ifdef OPENSSL_VERSION_PRE_RELEASE
```

The C preprocessor doesn't compare strings...

kroeckx commented on doc/man3/OPENSSL_VERSION_NUMBER.pod:
```diff
@@ -1,113 +0,0 @@
-=head1 NAME
-
-OPENSSL_VERSION_NUMBER, OPENSSL_VERSION_TEXT, OpenSSL_version,
-OpenSSL_version_num - get OpenSSL version number
```
Some of those are no longer documented.

kroeckx replied:
Or just failed to get mentioned in NAME / SYNOPSIS

levitte replied:
Yup

kroeckx commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,186 @@
+
+For an unknown B<t>, the text "not available" is returned.
+
+=head1 BACKWARD COMPATIBILITY
```
Please make it clear that they are deprecated

levitte replied:
Yes

kroeckx commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,186 @@
+OPENSSL_version_patch,
+OPENSSL_version_pre_release, OPENSSL_version_build_metadata
+OPENSSL_version_text, OPENSSL_version_text_full, OPENSSL_info
+- get OpenSSL version number
```
We should probably also document how to test that you're not using a deprecated function here. We can't generate warnings for all of them, for instance some are defines and we can't generate a warning when you're using a deprecated define, so you need to define something else that makes those deprecated defines go away.

levitte replied:
We have that when configuring

kroeckx commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,186 @@
+
+The macro B<OPENSSL_VERSION_PRE_RELEASE> is an added bit of text that,
+when defined, indicates that this is a pre-release version (possibly
+development, for which the value is "-dev").
```
Will this contain other data like for instance "-alpha2"?

levitte replied:
That's the idea, yeah

levitte commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,186 @@
+when defined, indicates that this is a pre-release version (possibly
+development, for which the value is "-dev").
+
+The macro B<OPENSSL_VERSION_BUILD_METADATA> is extra metadata, reserved
```
I don't really want to signal that it's reserved to *vendors* per se.  I should probably edit the docs...

richsalz replied:
I was thinking of a "vendor" as a "packager"  I think "metadata" is a bad/misleading term.  Vendor might not be as good as it should be, admittedly.

levitte replied:
I'm changing the docs to clarify this

tmshort replied:
Rather than "build_metadata", perhaps "suffix"?

levitte commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,186 @@
+
+=head2 Functions
+
+OpenSSL_version_num() returns the value of B<OPENSSL_VERSION_NUMBER>.
```
Will do

levitte committed:
util/find-doc-nits: allow #undef

levitte commented:
@t-j-h said thusly:
> A range of comments:
> Deprecation can only occur at a MAJOR version number - so the tests for 3_0 should be just 3.

Agreed...  when I wrote that branch and was thinking in terms of semantic versioning, I was thinking that we might want to support deprecating at any minor release, but since we're not going there, you're right. Will fix.

> We should always use the API-level testing macro and not raw check a synthesized number - i.e. the macro that gets passed MAJOR,MINOR should be what we use in general.

Yeah.  There were some spots where it wasn't quite clear to me what the intent was.  Will fix.

> If we need to test down at the MAJOR.MINOR.PATCH level we should have a macro for that purpose too - but raw 0x300000000 and 0x1010xxx should simply go from the code base so that we can in future change how we represent the version number in a single place. i.e. all usage of OPENSSL_API_COMPAT should be changed IMHO. We do not want users doing any of those sorts of tests in the future

Okie.  The reason I didn't touch `OPENSSL_API_COMPAT` is that there is a lot written in CHANGES on how to handle it, so I didn't want to break compatibility...  I will look at this.

> There are multiple places where you token-paste to get strings which are in C files which should be in public header files so that others can refer to them at compile-time too rather than runtime.
The macros should not use meta_stuff and pre_stuff as names (especially if they are present in a public header file)

So basically, for any macro that can be undefined, we should provide a variant that's guaranteed to be a string?  Something like this?

``` C
#ifdef OPENSSL_VERSION_PRE_RELEASE
# define OPENSSL_VERSION_PRE_RELEASE_STR OPENSSL_VERSION_PRE_RELEASE
#else
# define OPENSSL_VERSION_PRE_RELEASE_STR ""
#endif
```

> Could we change DEPRECATEDIN_X_Y(f) to DEPRECATEDIN(x,y,f) ???

No...  trying this would essentially mean trying to have CPP producing CPP code, and that's just not possible.

levitte committed:
Refactor the computation of API version limits

Previously, the API version limit was indicated with a numeric version
number.  This was "natural" in the pre-3.0.0 because the version was
this simple number.

With 3.0.0, the version is divided into three separate numbers, and
it's only the major number that counts, but we still need to be able
to support pre-3.0.0 version limits.

Therefore, we allow OPENSSL_API_COMPAT to be defined with a pre-3.0.0
style numeric version number or with a simple major number, i.e. can
be defined like this for any application:

    -D OPENSSL_API_COMPAT=0x10100000L
    -D OPENSSL_API_COMPAT=3

Since the pre-3.0.0 numerical version numbers are high, it's easy to
distinguish between a simple major number and a pre-3.0.0 numerical
version number and to thereby support both forms at the same time.

Internally, we define the following macros depending on the value of
OPENSSL_API_COMPAT:

    OPENSSL_API_0_9_8
    OPENSSL_API_1_0_0
    OPENSSL_API_1_1_0
    OPENSSL_API_3

They indicate that functions marked for deprecation in the
corresponding major release shall not be built if defined.

levitte commented:
I believe I've addressed all expressed concerns so far at this point.

levitte commented:
Remaining Travis's failing jobs seem to be caused by some Travis screwup (it says there's a .pod that shouldn't be there) in one case, and the missing fix from #7522 (I just commented on that)

vdukhovni commented:
It seems some folks are forgetting why we have the OPENSSL_API_COMPAT macros...

They make it possible for a user project to *incrementally* increase their base OpenSSL API, test disable all features deprecated prior to the oldest release they support, without breaking compatibility with later OpenSSL releases that deprecate even more features that the project may still be using.

A single boolean deprecated/not-deprecated fails to do provide what OPENSSL_API_COMPAT provides.

Furthermore, incompatible changes in a just-introduced compatibility interface are not a good idea.  We can introduce new macros that make it easier to construct the floor values, but please DO NOT remove the current mechanism controlled deprecation.

levitte commented:
Uhmmmm, I do believe the code in `openssl conf.h.in` keeps the incremental nature of `OPENSSL_API_COMPAT`.

t-j-h commented on include/openssl/opensslconf.h.in:
```diff
@@ -52,9 +52,14 @@ extern "C" {
+ * For any version number up until version 1.1.x, <version> is expected to be
+ * the calculated version number 0xMNNFFPPSL.  For version numbers 3.0.0 and
+ * on, <version> is expected to be only the major version number (i.e. 3 for
+ * version 3.0.0).
```
It should be in the pod docs and also in the header file - and consistent.

levitte replied:
For build time, it's documented in `INSTALL`, see:

https://github.com/openssl/openssl/blob/603221407ddc6404f8c417c6beadebf84449074c/INSTALL#L134-L138

and:

https://github.com/openssl/openssl/blob/603221407ddc6404f8c417c6beadebf84449074c/INSTALL#L320-L323

As for documenting `OPENSSL_API_COMPAT`, I'm all for it.  Question is where.  Any suggestion?  Should we have a feature macro pod?

levitte replied:
Note: `OPENSSL_API_COMPAT` is documented quite thoroughly in `CHANGES`...

mspncp replied:
The manual pages are the first place where the developer seeks information, so this is the place where he should find the relevant information.

levitte replied:
I'm starting to think that we should have a page that describes supported user-defined macros, a little like [feature_test_macros(7)](http://man7.org/linux/man-pages/man7/feature_test_macros.7.html), which can be referred to from other places.

mspncp replied:
Sounds good.

levitte replied:
Okie.  I think that's subject for another PR that collects all those macros.  Are y'all ok with that?  If you are, I'll consider this conversation resolved.

levitte commented on util/libcrypto.num:
```diff
@@ -261,7 +261,7 @@ ASN1_NULL_free                          262	1_1_0	EXIST::FUNCTION:
 EC_KEY_copy                             263	1_1_0	EXIST::FUNCTION:EC
 EVP_des_ede3                            264	1_1_0	EXIST::FUNCTION:DES
```
You're right.  Will fix

levitte replied:
I realised that this could be done automatically with a base version in `OpenSSL::Ordinals`.  See #7740 for my solution.

levitte commented:
> Remaining Travis's failing jobs seem to be caused by some Travis screwup (it says there's a .pod that shouldn't be there) in one case, and the missing fix from #7522 (I just commented on that there)

The last bit got fixed by #7738, which I just merged, and then rebase this PR on top of.  Travis should become a *little* happier.

mspncp commented:
Hmm... [Travis Job 462359821](https://travis-ci.org/openssl/openssl/jobs/462359821) exited with error code 1, but I don't see any cause for the failure?

mspncp commented:
Ah, doc-nits failed:

```
5.12s$ if [ -n "$CHECKDOCS" ]; then if $make doc-nits; then echo -e '+\057\057 MAKE DOC-NITS OK'; else echo -e '+\057\057 MAKE DOC-NITS FAILED'; false; fi; fi
(cd .; /usr/bin/perl util/find-doc-nits -n -p ) >doc-nits
doc/man3/OPENSSL_version_text.pod:1: the following exist as other .pod files:
OPENSSL_VERSION_TEXT
make: *** [doc-nits] Error 1
+// MAKE DOC-NITS FAILED
The command "if [ -n "$CHECKDOCS" ]; then if $make doc-nits; then echo -e '+// MAKE DOC-NITS OK'; else echo -e '+// MAKE DOC-NITS FAILED'; false; fi; fi" exited with 1.
```

vdukhovni commented:
> @levitte ... have you had time to verify that the incremental nature of OPENSSL_API_COMPAT is preserved, ...

No, not yet.  Nor was my earlier comment suggesting that anything was broken, I was just reacting a snippet of the comments made in this thread, without yet having seen the underlying changes.  It seems that the compatibility framework is being preserved, or at least is not intended to be broken, so that's great.

Presently, I am busy adding SNI support to Postfix, and finding some friction that really does need to be addressed, I hope to attract some attention to #7743 even with all the FIPs-related work going on.  SNI should not be that difficult. I am willing to contribute some of the implementation work once we decide what should be done, but I also need guidance on what can be with 1.0.2, in which SNI support will not change at this point, beyond perhaps better documentation...

levitte commented:
> No, not yet. Nor was my earlier comment suggesting that anything was broken, I was just reacting a snippet of the comments made in this thread, without yet having seen the underlying changes. It seems that the compatibility framework is being preserved, or at least is not intended to be broken, so that's great.

Message received, I will have a look at the comments.  I admit to not having given them my full attention.

> Presently, I am busy adding SNI support to Postfix, and finding some friction that really does need to be addressed, I hope to attract some attention to #7743 even with all the FIPs-related work going on. SNI should not be that difficult. I am willing to contribute some of the implementation work once we decide what should be done, but I also need guidance on what can be with 1.0.2, in which SNI support will not change at this point, beyond perhaps better documentation...

I'll try to have a look...  libssl has never been my forte, though.

levitte commented:
Also, you might be interested in #7742, where I specifically try to document `OPENSSL_API_COMPAT`

levitte commented:
I wonder, is this possibly a comment that made you wonder?

```
 * For any version number up until version 1.1.x, <version> is expected to be
 * the calculated version number 0xMNNFFPPSL.  For version numbers 3.0.0 and
 * on, <version> is expected to be only the major version number (i.e. 3 for
 * version 3.0.0).
```

The reason for going with major versions only is that there's a policy decision to *only introduce new deprecations at new major release boundary*, so limiting the future of `OPENSSL_API_COMPAT` on major version number only should be incremental enough.  Note that the code still supports the `0xN0000000L` style for a foreseeable future (at least until `N == 0x7`, which I expect to be quite far in the future :wink:).

I hope that's conclusive enough for an approval at this point.

levitte commented:
Rebased and with modifications to `doc/man7/openssl_user_macros.pod.in`

tmshort commented:
```
 * For any version number up until version 1.1.x, <version> is expected to be
 * the calculated version number 0xMNNFFPPSL.  For version numbers 3.0.0 and
 * on, <version> is expected to be only the major version number (i.e. 3 for
 * version 3.0.0).
```
This issue I see with this is that new APIs may be introduced at almost any time (I'm thinking of minor version increments), and this would prevent anyone from checking for them. I think the minor version number should be included as well.

levitte commented:
> This issue I see with this is that new APIs may be introduced at almost any time (I'm thinking of minor version increments), and this would prevent anyone from checking for them. I think the minor version number should be included as well.

This is a *whole* other can of worms, as we have no marking at all to support this in the header files.  This is about compatibility backward, right?  I.e. check if source built around OpenSSL 1.1.1 can also be compiled against 1.1.0?  We have no special support for this, and introducing that is a *lot* of work.

The `OPENSSL_API_COMPAT` macro was introduced to incrementally check how far forward one can build some piece of software without getting errors for missing (deprecated) symbols.

levitte commented:
Note that I'm not trying to dismiss the backward check off-handedly.  I just think that it doesn't belong in this PR.

tmshort commented:
But for this PR, I think:
```
 * For any version number up until version 1.1.x, <version> is expected to be
 * the calculated version number 0xMNNFFPPSL.  For version numbers 3.0.0 and
 * on, <version> is expected to be only the major version number (i.e. 3 for
 * version 3.0.0).
```

Should be:
```
 * For any version number up until version 1.1.x, <version> is expected to be
 * the calculated version number 0xMNNFFPPSL.  For version numbers 3.0.0 and
 * on, <version> is expected to be just the major and minor version number 
 * (i.e. 3.0 for version 3.0.0).
```
It's a change to just the comment for this PR, and only a commitment to update the minor version too (which is not done all that frequently). The more information that can be provided, the better for the end-users.

tmshort commented:
Looking through the diff a bit more carefully, I realize that it's not quite as simple as I had presumed, but I still think the minor version is important, and if it is to be included, should be done in this PR.

levitte commented:
> Basically, am I compiling against a version of OpenSSL that contains this new API yet?

Welllll....  if your compilation doesn't go through, it's quite obvious you aren't...

> That being said, I think the documentation should be updated (but not in this PR!) with the minimum OpenSSL version that an API was introduced. 

If you look at certain manuals, you will see that there's a HISTORY section...  we've been adding those as we go, more or less since we started working on the 1.1.0 release.

(BTW, should you find the time to dig into history and add a few lines to some manual you happen to look at, *PRs welcome*!!!  This can't be repeated enough)

> Looking through the diff a bit more carefully, I realize that it's not quite as simple as I had presumed, but I still think the minor version is important, and if it is to be included, should be done in this PR.

For the current usage, the minor version has *no* meaning whatsoever, because it's our policy to only introduce deprecations at major release boundary, and our major release will be tied (again, finally) with the major version number starting with 3.0.0.

tmshort commented:
>>Basically, am I compiling against a version of OpenSSL that contains this new API yet?
>>
>Welllll.... if your compilation doesn't go through, it's quite obvious you aren't...

I'm thinking of open-source programs that can take advantage of new OpenSSL features if available, rather than failing outright. Not dissimilar to programs supporting both OpenSSL 1.0.x/1.1.0.

>If you look at certain manuals, you will see that there's a HISTORY section... we've been adding those as we go, more or less since we started working on the 1.1.0 release.

Yes, and it would be better if the value to be checked against were in the documentation; rather than having to dig for it... but again, too much for this PR.

>For the current usage, the minor version has no meaning whatsoever, because it's our policy to only introduce deprecations at major release boundary, and our major release will be tied (again, finally) with the major version number starting with 3.0.0.

I'm not talking about deprecations... I'm talking about *new* APIs: function *Xyz()* added in 3.1, but not available in 3.0.

>So, I'm quite frankly not very keen on supporting minor numbers... yet, I'm considering it.

Thank you, the more information we can give the user, the better.

levitte commented:
> I'm not talking about deprecations...

[sigh] I *get* that.  However, `OPENSSL_API_COMPAT` is designed for deprecation checks.
So when I think of this, it needs a redesign to support what you want, and I think that's too big for this PR.

tmshort commented:
If `OPENSSL_VERSION_NUMBER` remains consistent, then it should allow for the appropriate checks (since that's what it seems most apps are using to check for things, now). But I'm not sure `OPENSSL_VERSION_NUMBER` can really ever be deprecated

levitte commented:
> But I'm not sure `OPENSSL_VERSION_NUMBER` can really ever be deprecated

I'm pretty sure that by version 8.0.0, it will stop working correctly.  It should be *removed* by then.

levitte commented:
Is there anything more blocking this?

mspncp commented on CHANGES:
```diff
@@ -7,7 +7,23 @@
+ Changes between 1.1.1 and 3.0.0 [xx XXX xxxx]
+
+  *) Change the possible version information given with OPENSSL_API_COMPAT.
+     It may be a pre-3.0.0 style numerical version number as was defined in
```
as _it_ was (?)

levitte replied:
Sure

mspncp commented on CHANGES:
```diff
@@ -7,7 +7,23 @@
+     It may be a pre-3.0.0 style numerical version number as was defined in
+     1.1.0, and it may also simply take the major version number.
+
+     Because of the practiced version numbering of pre-3.0.0 releases, the
```
Because of the version numbering used by pre-3.0.0 releases, ...

levitte replied:
Okie

mspncp reviewed (commented):
Looks good from a bird's eye perspective (i.e without digging too much into perl details). Only a few nonbinding suggestions.

Apart from that, due to its importance I believe this should be approved by another @openssl/omc, not a committer.

mspncp commented on CHANGES:
```diff
@@ -7,7 +7,23 @@
+     changes may introduce incompatible API/ABI changes.  MINOR number
+     changes may introduce new features and retain API/ABI compatibility.
+     PATCH number changes are for bug fixes and similar and retain API/ABI
+     compatibility.
```
One more suggestion:
```
 *) Switch to a new version scheme using three numbers MAJOR.MINOR.PATCH. 
 
    * Major releases (indicated by incrementing the MAJOR release number)
      may introduce incompatible API/ABI changes.
    * Minor releases (indicated by incrementing the MINOR release number)
      may introduce new features but retain API/ABI compatibility.
    * Patch releases (indicated by incrementing the PATCH number)
      are intended for bug fixes and other improvements of existing
      features only (like improving performance or adding documentation)  
      and retain API/ABI compatibility.
```

levitte replied:
Yup!

mspncp commented:
Fixups look good.

levitte commented:
@mspncp requested that another @openssl/omc member review and rubber stamp this.  I think that's a good idea in this case.

mattcaswell commented:
doc-nits failure reported by travis:

````
$ if [ -n "$CHECKDOCS" ]; then if $make doc-nits; then echo -e '+\057\057 MAKE DOC-NITS OK'; else echo -e '+\057\057 MAKE DOC-NITS FAILED'; false; fi; fi

(cd .; /usr/bin/perl util/find-doc-nits -n -p ) >doc-nits

doc/man3/OPENSSL_version_text.pod:1: the following exist as other .pod files:

OPENSSL_VERSION_TEXT

make: *** [doc-nits] Error 1

+// MAKE DOC-NITS FAILED
````

mattcaswell commented on include/openssl/opensslv.h:
```diff
@@ -10,92 +10,144 @@
+#  ifdef OPENSSL_NO_NEWVERSION_ONLY
+/* Helper functions */
+#   define OPENSSL_MSTR_HELPER(x) #x
+#   define OPENSSL_MSTR(x) OPENSSL_MSTR_HELPER(x)
```
I'm confused by the above line. Can you explain why we need it?

levitte replied:
Classic stringify of C preprocessor macro values.  See here: https://stackoverflow.com/questions/2653214/stringification-of-a-macro-value

mspncp replied:
TL;DR: the `#` operator doesn't expand its argument by design. If you want expansion to happen, you need two nested macro calls, the first of which expands its arguments and the second one (the helper) does the string conversion.

mattcaswell commented on include/openssl/opensslv.h:
```diff
@@ -10,92 +10,144 @@
+    (long)( (OPENSSL_VERSION_MAJOR<<28)  \
+            |(OPENSSL_VERSION_MINOR<<20) \
+            |(OPENSSL_VERSION_PATCH<<4)  \
+            |0x0 )
```
Should this last digit have a different value depending on whether `OPENSSL_VERSION_PRE_RELEASE` is defined? i..e if `OPENSSL_VERSION_PRE_RELEASE` is defined then it has the value `0x0`, otherwise it has the value `0xf`

levitte replied:
I dunno.  I was hoping that would not be significant, but maybe it is.  What's your opinion?

levitte replied:
Okie

mattcaswell commented on include/openssl/pkcs12.h:
```diff
@@ -55,7 +55,7 @@ typedef struct pkcs12_bag_st PKCS12_BAGS;
 /* Compatibility macros */
 
-#if OPENSSL_API_COMPAT < 0x10100000L
+#ifndef OPENSSL_API_1_1_0
```
Do all the header files where you make similar changes ultimately include opensslconf.h? If not applications that define OPENSSL_API_COMPAT may not pick up the definitions of OPENSSL_API_1_1_0 etc and hence start declaring things that shouldn't be declared.

levitte replied:
This is a good point, and hard to detect as written.  So maybe those macros should be turned into booleans instead, and checked with `#if` rather than `#ifdef`, that would make for better detection.

mattcaswell replied:
I'm not sure I understand how that helps?

levitte replied:
By making these macros boolean, we *demand* that they be defined, which *demands* that `opensslconf.h` be included first.  Our own builds will fail if this isn't the case...  if you want this checked even more thoroughly, it wouldn't be too hard to generate a build test program with all deprecated symbols and test that it builds properly.

mattcaswell commented on test/recipes/80-test_cipherlist.t:
```diff
@@ -20,7 +20,7 @@ setup("test_cipherlist");
 plan skip_all =>
     "This test recipe isn't supported when doing regression testing"
-    if $build_version != $library_version;
+    if $build_version ne $library_version;
```
Is this a bug fix?

levitte replied:
Yes.  `test/versions.c` doesn't print the hex versions any more, so neither does `OpenSSL::Test::openssl_versions()`.

mattcaswell commented on util/find-doc-nits:
```diff
@@ -115,7 +115,7 @@ sub name_synopsis()
             # an enumeration: enum ... {
             $sym = $1;
-        } elsif ( $line =~ /#define ([A-Za-z0-9_]+)/ ) {
+        } elsif ( $line =~ /#(?:define|undef) ([A-Za-z0-9_]+)/ ) {
```
Why is this needed?

levitte replied:
Because `doc/man3/OPENSSL_version_text.pod` has a `#undef` line in the synopsis, and there's no reason why that can't there.

levitte commented:
> doc-nits failure reported by travis:

That has already been commented upon [here](https://github.com/openssl/openssl/pull/7724#issuecomment-443362823), [here](https://github.com/openssl/openssl/pull/7724#issuecomment-443506711) (with followups).  It looks like a Travis screwup.

levitte commented on Configure:
```diff
@@ -43,8 +43,9 @@ my $usage="Usage: Configure [no-<cipher> ...] [enable-<cipher> ...] [-Dxxx] [-lx
-#               interfaces deprecated as of the specified OpenSSL version.
+# --api         One of 0.9.8, 1.0.0, 1.1.0 or 3.0.0 (or 3).  Do not compile
+#               support for interfaces deprecated as of the specified OpenSSL
+#               version.
```
Should that happen in this PR or in a different one?

levitte replied:
Ok, will do

mattcaswell replied:
This was marked as resolved but it doesn't seem to have been?

mattcaswell replied:
Ah...I think the underlying code has been changed but the docs here haven't been updated.

levitte replied:
Ok, fixing

levitte commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,204 @@
+ #define OPENSSL_VERSION_MINOR  y
+ #define OPENSSL_VERSION_PATCH  z
+
+ #define OPENSSL_VERSION_PRE_RELEASE "-dev"
```
Should I add a comment saying that it's just an example?  I was thinking that the description would be enough explanation, but perhaps not...

levitte replied:
Sure, will do

levitte commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,204 @@
+
+OpenSSL_version_num() returns the value of B<OPENSSL_VERSION_NUMBER>.
+
+OpenSSL_version() returns the same values as OPENSSL_info() for the same
```
The motivation is that `OpenSSL_version()` probably was just about the version number and not about all the other information it provides, and since `OPENSSL_info()` doesn't serve the version number at all (because it's not one calculated number any more), the old name doesn't correspond at all to what it actually does.

mattcaswell replied:
Well it provides information about this *version* of OpenSSL :-)

I can understand deprecating a function where the parameters or the return type is wrong. But deprecating and introducing a new version because the name isn't as descriptive as we would like kind of seems like breakage for breakage sake (ok we're not breaking anything yet since we're not removing the old function yet - but in time we will). It's not like the old name is actively misleading (as was the case in the rename of SSLv23_method to TLS_method).

levitte replied:
We disagree, I associate the name with what it returns, and I *do* find the name strongly misleading.

mattcaswell replied:
Ok. I raised it on project and asked for more input.

hyc replied:
So how do we get the version number in the future? Pretty much everything that uses OpenSSL expects to be able to find a simple numeric string somewhere. HTTP User-Agent strings and various other tools want it.

levitte replied:
`OPENSSL_version_text()` and `OPENSSL_version_text_full()`, which are documented in this very manual.

sam-github replied:
My two bits: The old name isn't great, but of the retrospectively not-great APIs in openssl that are a problem, this wouldn't even be on my list. I wouldn't delete the existing API.

Its not clear why this PR introduces a new `_info(t)` API that takes a number as an argument to select some bit of non-version information, but then introduces 7 new APIs to get various pieces of the openssl version. Why not just have an API per piece of information for version and non-version info? Of if less functions are an API goal, why not return a `const struct OpenSSL_info*`? That would be easy to get all the info out of, and could ABI-compatibly have things added to the end of the struct.

levitte replied:
5 of those new functions are to get the library's idea of the values for the corresponding 5 macros.  Remains two that could arguably be folded into returned strings by OPENSSL_info() or OpenSSL_version(), whichever we end up choosing, if that's more desirable.

I would rather not add a public structure...

vdukhovni replied:
The current `OpenSSL_version(what)`, which returns strings describing various features of the run-time OpenSSL library is a reasonably intuitive interface, it is mostly used just with `what == OPENSSL_VERSION`, and the name fits the primary use-case.

In Postfix for example, you can query that value via "postconf -T run-version":
```
$ uname -sr
FreeBSD 11.2-RELEASE-p4
$ postconf -T run-version
OpenSSL 1.1.1a  20 Nov 2018
```
```
$ uname -sr
Darwin 18.2.0
$ postconf -T run-version
LibreSSL 2.2.7
```

We just recently (in the API changes from 1.0.2 to 1.1.0) renamed the function from `SSLeay_version`, and changing it again IMHO just punishes the users who've gone to the trouble of adapting to the preferred 1.1.0 interface.  The API reporting the version should not change with every API version!

We did too much cosmetic renaming in the transition from 1.0.2 to 1.1.0, let's not repeat that mistake. API stability trumps cosmetic issues.

levitte replied:
I've heard you all and had a think, and I think I may have come up with a solution that doesn't clash all over the place.  Fixup commits coming up

levitte commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,204 @@
+=head1 HISTORY
+
+The macros and functions described here were added to OpenSSL 3.0.0,
+with the exception apart from the L</BACKWARD COMPATIBILITY> ones.
```
Ah, thanks.  I was probably interrupted right there and lost my train of thought...

levitte commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,204 @@
+
+=head1 COPYRIGHT
+
+Copyright 2000-2018 The OpenSSL Project Authors. All Rights Reserved.
```
It's a rewrite of `OPENSSL_VERSION_NUMBER.pod`, but you're right, the rewrite is so total that it is effectively a new file.  Will fix.

levitte commented on Configure:
```diff
@@ -374,6 +382,7 @@ my @disablables = (
     "scrypt",
     "sctp",
     "seed",
+    "newversion-only",
```
Will do

levitte commented on engines/e_padlock.c:
```diff
@@ -22,18 +22,9 @@
-#  endif
+# if !OPENSSL_VERSION_PREREQ(3,0)
+#  error "Only OpenSSL >= 3.0.0 is supported"
+# endif
```
Wellllll........  at one point, the ENGINE API was a bit more flexible in that regard.  However, I do agree that for built in engines, this is really silly.  However, the condition was already there, and I didn't have the guts to remove this entirely...  however, if you agree it should, I will.

mattcaswell replied:
I think it should go

levitte replied:
Okie

levitte commented on include/openssl/opensslconf.h.in:
```diff
@@ -66,53 +71,93 @@ extern "C" {
+ *
+ *      0.x.y   0       (0.9.8 was the last major release in this series)
+ *      1.0.x   1       (1.0.2 was the last major release in this series)
+ *      1.1.x   2       (1.1.0 was the last major release in this series)
```
D'oh, you're entirely correct!

mattcaswell commented on doc/man7/openssl_user_macros.pod.in:
```diff
@@ -40,6 +53,30 @@ meaningful:
+
+This form is a simple number that represents the major version number
+and is supported for version 3.0.0 and up.  For extra conventience,
+these numbers are also available:
```
I mean the change in the definition of `OPENSSL_API_COMPAT` specifically. We could choose to handle this separately to the `OPENSSL_VERSION` macros if we wanted. I guess my main concern is the ensure that applications that define the old OPENSSL_API_COMPAT format will still work - which is sort of related to the question I had about whether all the headers include opensslconf.h

levitte replied:
Is this code `opensslconf.h.in` good enough an answer?

https://github.com/openssl/openssl/blob/1ef17a85285155f09cbb85e92b68c341db196534/include/openssl/opensslconf.h.in#L100-L103

This is designed to work with the old form for *quite* some time, so people will have time (years, I suspect!) to change their habit to use just the major number instead of the hex code.

mattcaswell replied:
Ok

levitte commented on include/openssl/opensslconf.h.in:
```diff
@@ -52,9 +52,14 @@ extern "C" {
- * still won't see them if the library has been built to disable deprecated
+ * declarations of functions deprecated in or before <version>. Otherwise,
+ * they still won't see them if the library has been built to disable deprecated
  * functions.
```
Ok, fixed...  I hope.

levitte commented:
I think I've addressed everything but the `OpenSSL_version` question now.

t-j-h commented on doc/man3/OPENSSL_config.pod:
```diff
@@ -8,7 +8,7 @@ OPENSSL_config, OPENSSL_no_config - simple OpenSSL configuration functions
 
- #if OPENSSL_API_COMPAT < 0x10100000L
+ #ifndef OPENSSL_API_1_1_0
  void OPENSSL_config(const char *appname);
```
#if !
To be consistent

t-j-h commented on doc/man3/RAND_cleanup.pod:
```diff
@@ -8,7 +8,7 @@ RAND_cleanup - erase the PRNG state
  #include <openssl/rand.h>
 
- #if OPENSSL_API_COMPAT < 0x10100000L
+ #ifndef OPENSSL_API_1_1_0
```
#if !

FdaSilvaYY commented on doc/man3/OPENSSL_version_text.pod:
```diff
@@ -0,0 +1,193 @@
+alpha release, etc...
+The value must be a string.
+
+The macro B<OPENSSL_VERSION_BUILD_METADATA> is extra metadata, reserved
```
nit : is ***an*** extra

levitte replied:
No.  As far as I know, there's nothing called "a metadata" in English.  It's the same class of words as "people", you can't say "a people".

mspncp replied:
no, i disgree. I don’t know the precise grammatical term in english, but it’s not a countable amount. For example, you take „extra milk“ in your coffee, not „an extra milk“.

mspncp replied:
Too slow again...

t-j-h reviewed (approved):
Approved once the two nits are fixed and CIs are green.

FdaSilvaYY commented on doc/man7/openssl_user_macros.pod.in:
```diff
@@ -40,6 +53,30 @@ meaningful:
+=item C<m>
+
+This form is a simple number that represents the major version number
+and is supported for version 3.0.0 and up.  For extra conventience,
```
typo : For extra **conventience**

levitte replied:
Thank you.  Fixed

levitte commented:
@t-j-h, you were out of sync there, please look again.  The manuals still need fixing, though, that's entirely correct.

levitte commented:
Squashed all fixup commits

levitte commented:
@t-j-h, apropos CIs being green, it seems that one Travis job just won't.  See [previous comments on the matter](https://github.com/openssl/openssl/pull/7724#issuecomment-444515632) (I'm getting *really* tired of having to repeat this)

levitte commented:
Gah, there was another nit, I had forgotten to fix up `util/libcrypto.num`.  Is re-approval needed?

levitte commented:
Same with the manual after all the `OPENSSL_info` vs `OpenSSL_version` changes...

levitte commented:
Ooooh, all CIs green!

t-j-h reviewed (approved):
Green for go

mspncp commented:
What about @mattcaswell's red cross? Shouldn't it be dismissed at least?

levitte commented:
As far as I can tell, all of @mattcaswell's concerns have been answered at this point.  I'm not in the habit of dismissing reviews.  Besides, it seems the better habit is to resolve each review conversation individually...  which has been done.

mattcaswell commented on INSTALL:
```diff
@@ -393,6 +393,11 @@
+  newversion-only
+                   Don't build with pre-3.0.0 version scheme macros, i.e.
+                   remove support for the macros OPENSSL_VERSION_NUMBER and
+                   OPENSSL_VERSION_TEXT.
```
Why do we need a new option for this? Who will use this?

levitte replied:
Oh, that doc needs updating.

> Who will use this?

Anyone that wants to try with the new macros only.

mattcaswell replied:
I'm not sure why someone would want to do that. It doesn't seem like a compelling reason to add a new config option - we've got too many of those as it is. I think we should remove this.

levitte replied:
Doc updated

levitte replied:
Hmm, I guess that `no-deprecated` or `--api=3.0.0` should suffice, so ok

mattcaswell commented:
Looks ok other than the couple of comments I just added.

levitte committed:
fixup! Refactor the computation of API version limits

levitte committed:
fixup! Switch to MAJOR.MINOR.PATCH versioning and version 3.0.0-dev

levitte committed:
fixup! Switch to MAJOR.MINOR.PATCH versioning and version 3.0.0-dev

levitte commented:
Looks ok with the latest small fixups?

levitte commented:
Argh...  sorry :-/
